File: cpp/0091.cpp
#include <string>
#include <vector>
#include <iostream>

using namespace std;

class Solution {
public:
    int numDecodings(string s) {
        int n = s.size();
        if (s[0] == '0') return 0;
        if (n == 1) return 1;
        int curr = 1;
        int prev = 1;
        for (int i = 1; i < n; i++) {
            int next = 0;
            // one letter
            if (s[i] != '0') {
                next += curr;
            }
            // two letters
            if (s[i-1] == '1' || (s[i-1] == '2' && s[i] <= '6')) {
                next += prev;
            }
            prev = curr;
            curr = next;
        }
        return curr;
    }
};

int main() {
    Solution sol;
    cout << sol.numDecodings("12") << "\n";
    cout << sol.numDecodings("226") << "\n";
    cout << sol.numDecodings("0") << "\n";
    cout << sol.numDecodings("06") << "\n";
    return 0;
}

File: python/0091.py
# https://leetcode.com/problems/decode-ways/

class Solution:
    def numDecodings(self, s: str) -> int:
        N = len(s)
        if N == 0 or s[0] == '0': return 0
        curr, prev = 1, 1
        for i in range(1, N):
            count = 0
            if s[i] != '0':
                count += curr
            if 10 <= int(s[i-1:i+1]) <= 26:
                count += prev
            curr, prev = count, curr
        return curr


assert Solution().numDecodings('0') == 0
assert Solution().numDecodings('06') == 0
assert Solution().numDecodings('226') == 3
